### ISSUE TITLE
[Unity][MSC][Tracking Issue] Introduction to Multi-System Compiler

### ISSUE THREAD

**Issue by Archermmt** [apache/tvm#15233]
- [ ] [M0] Build MSCGraph core parts. Enable translation between Relay, Relax and MSCGraph without lossing information.
- [ ] [M1] Finish RuntimeManager for relax, and torch, so that a compiling process can be test based on MSCGraph.
- [ ] [M2] Use MSCProcessor to manage the compiling pipeline. Add pruner, prune the graph by given density.
- [ ] [M3] Add MSCGym, enable auto pruning process. Add distiller, enable knowledge distilliation for pruning.
- [ ] [M4] Add quantizer and collecter for quantization and debugging.
- [ ] [M5] [Optional] Add MSCWrapper as compression toolchain. MSCWrappers are wrapper format of MSCProcessor, which can be used to wrap and optimize the models in training system.
- [ ] [M6] [Optional] Add MSC_PIPE for Saas. Build a local service and client for test the Saas

**Comment by Archermmt** [comment]
Intorduction @ https://discuss.tvm.apache.org/t/rfc-unity-msc-introduction-to-multi-system-compiler/15251

**Comment by Archermmt** [comment]
TODO: add tests for M0.2 after M0.3

**Comment by Archermmt** [comment]
Discussion on translate relay to relax without loss info: https://discuss.tvm.apache.org/t/msc-translate-relay-to-relax-without-loss-info/15650

**Comment by Lunderberg** [comment]
I'm somewhat concerned about the `relay -> python codegen -> relax` code path used in `tvm.contrib.msc.framework.torch.frontend.translate.from_torch` when `via_relax=False`.  This is a duplication of the serialization/parsing used in TVMScript (`tvm.script`), and can cause CI failures (e.g https://github.com/apache/tvm/pull/15783) due to this duplication.

While I agree with the need for a operator-level conversion from relay to relax, I think it should be done through extending the existing `relax.testing.relay_translator.from_relay` converter rather than having an additional python code-generator.

**Comment by Archermmt** [comment]
@Lunderberg  sorry for the late reply.... I've checkout the failures, which is cause by from_fx function of relax. The failures of test_translate_relax.py and test_translate_torch.py have the same reason. I think MSC is not the source of the failures, it just build new modules based on relax and relay, without touching the basic codes of them.

And the reasons way build a duplicate "relay -> relax" method:
1.  A operator-level conversion is needed, as you said. This is essential when developers want to use relay based features (like me, testing tensorflow).
2. Using relay also have some problems in optimizing the model, especially in quantization, pruning, parameter reusing and training. The real process is relay -> MSCGraph -> relax, MSCGraph is the key DAG structure in model compression. This via_relax=False only shows an example of using MSC with relax and relay, not meaning to be a converted between relay and relax. When the final solution for the "operator-level conversion from relay to relax" is done, I will change the relay-relax method accordingly.

Thanks for watching !

**Comment by Lunderberg** [comment]
@Archermmt No worries, and I've been slow responding as well.

After thinking on it, I think my primary concern is in the method used for the `MSCGraph -> relax` conversion, which is done by first producing a python string, then calling `exec` on the generated string.  This makes it very difficult to tell where an error has been introduced, as any errors in this process are thrown at runtime while executing the generated string.

Instead of generating a string to use the Python API, I think the MSC to Relax conversion should instead be done by directly calling the C++ APIs.  This would expose any errors during the C++ compilation, rather than delaying them until runtime.

**Comment by Archermmt** [comment]
@Lunderberg Emmm....I've also thought about this, which method is better: 1. Convert in C++ to enable eager errors detection; 2. Convert by string generation to enable independent loading. Both has advantage and disadvantage.

The first method (lets say converter, either C++ or python) like relax.builder can check and normalize the op while building graph, but that limit the deployment possibility. For example if I need compare the results between an old version tvm without relax and the new unity version(which maybe a real task for me....), I have to spend lot of time setting up environments and dumps testing datas with the converter solution. And MSC is designed not only for converting to relax, but also torch/torch2, tensorflow/tf2, tensorrt, and so on. Considering dispatch models in different framework and environment, the converter may not be a good solution.

The second method (lets say string generation) like cutlass codegen first generate strings and process them to kernel/model/engine. That means codegen process disable check and normalization, that may lead to lazy errors detection. However, strings can be change to script/C++ files and loaded in any environment, that method seperates codegen and loading, which is very essential in fast model release, especially on cloud(where different environment and framework are used).

And as mentioned in the RFC:https://discuss.tvm.apache.org/t/rfc-unity-msc-introduction-to-multi-system-compiler/15251
MSC is currently targeting at solving the model optimization problems base on relax. That means the codegen part should have the ability of using features in different framework, such as training, weights reusing/reloading, distribution system, and so on. Current I only have experience "describe" these features in python with string generation(not that good at C++ -_-).

To partially solve the error detection problem, the codegen in MSC not only generate the model, but also generate the unittest. Using the unittest developers can locate and solve the problems efficiently.

I think we can leave this part as a todo, thus enable C++ converter for MSC. After the main target is reached, I'll consider of building a converter, or may be directly use relax as the core IR.
